src/ledger/recovery.zig

```diff
@@ -0,0 +1,500 @@
+            common_header.index = try checkedSub(common_header.index, position);
+            code_header.position = 0;
+            break .{
+                .common_header = common_header,
+                .code_header = code_header,
+                .chained_merkle_root = chained_merkle_root,
+                .retransmitter_signature = retransmitter_signature,
+            };
+        }
+    } else return error.TooFewParityShards;
+    std.debug.assert(verifyErasureBatch(meta.common_header, meta.code_header, shreds));
+    std.debug.assert(!meta.common_header.variant.resigned or
+        meta.retransmitter_signature != null);
+    return meta;
+}
+
+fn organizeShredsForRecovery(
+    allocator: Allocator,
+    shreds: []const Shred,
+    meta: RecoveryMetadata,
+) !RecoveryShreds {
+    const num_shards: usize =
+        @intCast(meta.code_header.num_data_shreds + meta.code_header.num_code_shreds);
+
+    // Obtain erasure encoded shards from shreds.
+    const input_shreds = try allocator.alloc(?Shred, num_shards);
+    errdefer allocator.free(input_shreds);
+    const shards = try allocator.alloc(?[]const u8, num_shards);
+    errdefer {
+        for (shards) |shard| if (shard) |s| allocator.free(s);
+        allocator.free(shards);
+    }
+
+    for (input_shreds) |*s| s.* = null;
+    for (shards) |*s| s.* = null;
+    for (shreds) |shred| {
+        const index = shred.erasureShardIndex() catch {
+            return error.InvalidIndex;
+        };
+        if (index >= shards.len) return error.InvalidIndex;
+        input_shreds[index] = shred;
+        shards[index] = try shred.erasureShardAsSlice();
+    }
+
+    // identify which shreds were already present and which need to be recovered
+    var num_to_recover: usize = 1;
+    const mask = try allocator.alloc(bool, input_shreds.len);
+    errdefer allocator.free(mask);
+    for (input_shreds, 0..) |s, i| {
+        mask[i] = s != null;
+        if (s == null) {
+            num_to_recover += 1;
+        }
+    }
+
+    return .{
+        .allocator = allocator,
+        .input_shreds = input_shreds,
+        .shards = shards,
+        .mask = mask,
+        .num_to_recover = num_to_recover,
+    };
+}
+
+/// Reconstruct code and data shreds from erasure encoded shards.
+fn reconstructShreds(
+    allocator: Allocator,
+    meta: RecoveryMetadata,
+    shreds: RecoveryShreds,
+) ![]const Shred {
+    // Reconstruct code and data shreds from erasure encoded shards.
+    const all_shreds = try allocator.alloc(Shred, shreds.input_shreds.len);
+    var num_recovered_so_far: usize = 0;
+    errdefer {
+        // The shreds that are created below need to be freed if there is an error.
+        // If no error, ownership is transfered to calling scope.
+        // Existing shreds do not need to be freed because they were already owned by calling scope.
+        for (all_shreds, shreds.mask, 0..num_recovered_so_far) |shred, was_present, _| {
```

Comment:
multi-for loops in Zig assert that all elements are the same length. The `0..num_recovered_so_far` doesn't do anything.

Reply:
> The 0..num_recovered_so_far doesn't do anything.

It does do something. It triggers a panic if an error happens during shred initialization. Which is obviously not what we want. I assumed the shortest one would constrain the loop, since unbounded ranges are always constrained by a bounded range. Guess I didn't know how loops work.

73b26d10c514dfb80ae8b65f34076638c2b022ca

Reply:
35763cb991c8f8894e5f915d6cecf3c8741e79fe

Reply:
I don't agree with the `ArrayList` solution, it's unnecessary overhead. If you need to constrain to the shortest one do something to the effect of:
```zig
const min_length = @min(all_shreds.len, shreds.mask.len, num_recovered_so_far);
for (all_shreds[0..min_length], shreds.mask[0..min_length]) ...
```
This 1. better describes what the code is trying to do and 2. Is just more efficient than an `ArrayList`.

Reply:
How is this more efficient? Are you talking about the extra logic in `ensureTotalCapacityPrecise`?

I knew I could accomplish this without a list. The sole reason I chose to use a list is because I feel that it actually better describes what the code is trying to do.

When you iterate over the list, you're only going to access items that were added to the list. That's ultimately the goal here. It's also the first assumption I would expect to have about how a list would behave here. It's nice when the behavior of code is intuitive and simple to describe. It's also slightly more concise.

With the slice allocation there is this situation where the slice length actually includes undefined regions of memory and you need to carefully track it with a separate variable to avoid undefined behavior. It's a footgun that may not be obvious to every reader. The code to deal with it doesn't really help understand what the function is trying to accomplish. I find it cleaner to use a data structure that doesn't expose the footgun in the first place.

Ultimately it's not really a big deal, but I just wanted to express why I switched to using a list. But if there's a meaningful performance penalty then I could see removing the list.

Reply:
I understand why you're using a list here. Those seem like good reasons to me, and I'm all about showing code intent. I just need to think of the code as "accessing how many items were added" and not "accessing the smallest range possible", and it makes more sense.

> But if there's a meaningful performance penalty then I could see removing the list.

Probably not, but without a good reason (before you explained it to me), I would have rather just used the faster option.


Anyways, makes sense, I'm happy with it.